http-common/src/connector.rs

```diff
@@ -0,0 +1,335 @@
+		loop {
+			let stream = match &mut *self {
+				Incoming::Http(listener) => match listener.poll_accept(cx) {
+					std::task::Poll::Ready(Ok((stream, _))) => Ok(AsyncStream::Http(stream)),
+					std::task::Poll::Ready(Err(err)) => Err(err),
+					std::task::Poll::Pending => return std::task::Poll::Pending,
+				},
+
+				Incoming::Unix(listener) => {
+					// tokio::net::UnixListener does not have pub poll_accept.
+					//
+					// However the tokio::net::unix::Incoming returned from its .incoming() does.
+					// Keeping an Incoming across polls is hard because it borrows a &mut of the UnixListener.
+					// But it's fine to throw it away when it's Pending and make a new one for every poll, because it doesn't contain any state
+					// nor has side effects from being dropped.
+					let mut incoming = listener.incoming();
+					match std::pin::Pin::new(&mut incoming).poll_accept(cx) {
+						std::task::Poll::Ready(Ok(stream)) => Ok(AsyncStream::Unix(stream)),
+						std::task::Poll::Ready(Err(err)) => Err(err),
+						std::task::Poll::Pending => return std::task::Poll::Pending,
+					}
+				},
+			};
+
+			match stream {
+				Ok(stream) => return std::task::Poll::Ready(Some(Ok(stream))),
+				Err(err) => match err.kind() {
+					// Client errors
+					std::io::ErrorKind::ConnectionAborted |
+					std::io::ErrorKind::ConnectionRefused |
+					std::io::ErrorKind::ConnectionReset => (),
+
+					_ => return std::task::Poll::Ready(Some(Err(err))),
+				},
+			}
+		}
+	}
+}
+
+impl hyper::service::Service<hyper::Uri> for Connector {
+	type Response = AsyncStream;
+	type Error = std::io::Error;
+	type Future = std::pin::Pin<Box<dyn std::future::Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+	fn poll_ready(&mut self, _cx: &mut std::task::Context<'_>) -> std::task::Poll<Result<(), Self::Error>> {
+		std::task::Poll::Ready(Ok(()))
+	}
+
+	fn call(&mut self, _req: hyper::Uri) -> Self::Future {
+		match self {
+			Connector::Http { host, port } => {
+				let (host, port) = (host.clone(), *port);
+				let f = async move {
+					let inner = tokio::net::TcpStream::connect((&*host, port)).await?;
+					Ok(AsyncStream::Http(inner))
+				};
+				Box::pin(f)
+			},
+
+			Connector::Unix { socket_path } => {
+				let socket_path = socket_path.clone();
+				let f = async move {
+					let inner = tokio::net::UnixStream::connect(&*socket_path).await?;
+					Ok(AsyncStream::Unix(inner))
+				};
+				Box::pin(f)
+			},
+		}
+	}
+}
+
+impl<'de> serde::Deserialize<'de> for Connector {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: serde::de::Deserializer<'de> {
+		struct Visitor;
+
+		impl<'de> serde::de::Visitor<'de> for Visitor {
+			type Value = Connector;
+
+			fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+				formatter.write_str("Connector")
```

Comment:
According to `serde::de::Visitor` dev docs - 
> The message should not be capitalized and
    should not end with a period

Reply:
Right, I misremembered what `#[derive(serde::Deserialize)]` expands to. It should be `"struct Connector"`. But anyway, `"an endpoint URI"` is better. Fixed.